### PR TITLE
Initialize feature flags before extensions manager

### PR DESCRIPTION
**Pull request by cwperks**
Signed-off-by: Craig Perkins <cwperx@amazon.com>

### Description

While testing with the experimental extensions feature flag, I found that the NoopExtensionsManager was being instantiated because the feature flags were not initialized by the time that extensions manager is being instantiated. This is a small PR to re-order Node.java to ensure the FeatureFlags are initialized before ExtensionsManager instantiation.

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwperks** [comment]
@ryanbogan @saratvemulapalli @dbwiddis Small PR on the order of initializing feature flags before instantiating the ExtensionsManager.

**Comment by cwperks** [comment]
@ryanbogan Is this a flaky test failure? Can the gradle check be re-run?

```
Tests with failures:
 - org.opensearch.indices.replication.SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -67,6 +67,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Fix 'org.apache.hc.core5.http.ParseException: Invalid protocol version' under JDK 16+ ([#4827](https://github.com/opensearch-project/OpenSearch/pull/4827))
 - Fixed compression support for h2c protocol ([#4944](https://github.com/opensearch-project/OpenSearch/pull/4944))
 - Support OpenSSL Provider with default Netty allocator ([#5460](https://github.com/opensearch-project/OpenSearch/pull/5460))
+- Initialize feature flags before extensions manager ([#6082](https://github.com/opensearch-project/OpenSearch/pull/6082))
```
Comment (andrross) [inline CHANGELOG.md:70]: Can probably [skip the changelog](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#which-changes-require-a-changelog-entry) on this one as it doesn't really impact user-facing functionality.

Comment (cwperks) [reply CHANGELOG.md]: Removed from the CHANGELOG

**Comment by ryanbogan** [comment]
> @ryanbogan Is this a flaky test failure? Can the gradle check be re-run?
> 
> ```
> Tests with failures:
>  - org.opensearch.indices.replication.SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh
> ```

I think that one is flaky

**Review by saratvemulapalli — APPROVED** [review]

**Comment by saratvemulapalli** [comment]
Thanks @cwperks !!

Ideally a test should have caught this problem. @ryanbogan / @cwperks do you want to contribute a test for `opensearch.yml` having feature flag set and expecting `ExtensionsManager` to run?

**Review by andrross — APPROVED** [review]

**Comment by cwperks** [comment]
@saratvemulapalli Are there any example ITs for extensions? I am looking into a setup similar to this: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java#L57-L64

Do you think its possible to use `internalCluster().getInstance(ExtensionManager.class)` to get the extension manager that was instantiated to see if its a NoopExtensionManager vs an ExtensionManager?

**Comment by saratvemulapalli** [comment]
> @saratvemulapalli Are there any example ITs for extensions? I am looking into a setup similar to this: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java#L57-L64
> 
> Do you think its possible to use `internalCluster().getInstance(ExtensionManager.class)` to get the extension manager that was instantiated to see if its a NoopExtensionManager vs an ExtensionManager?

Thats a good question. It should be possible to do it as we inject it[1]. 

[1] https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L1010

**Comment by cwperks** [comment]
@saratvemulapalli @ryanbogan @andrross I added a test for this. When adding the test, I had to remove the feature flag check around binding the extensions manager otherwise the call to `internalCluster().getInstance(ExtensionsManager.class)` would always return `ExtensionsManager` and not the actual instance that was bound to the node. Please let me know if there is a better way to do this. 

This means that when the experimental feature flag is not enabled, there will always be a `NoopExtensionsManager` that is binded. I figured that may alright, but want to double-check with the maintainers.

**Review by saratvemulapalli — APPROVED** [review]
Thanks @cwperks !!

**Comment by saratvemulapalli** [comment]
> @saratvemulapalli @ryanbogan @andrross I added a test for this. When adding the test, I had to remove the feature flag check around binding the extensions manager otherwise the call to `internalCluster().getInstance(ExtensionsManager.class)` would always return `ExtensionsManager` and not the actual instance that was bound to the node. Please let me know if there is a better way to do this.
> 
> This means that when the experimental feature flag is not enabled, there will always be a `NoopExtensionsManager` that is binded. I figured that may alright, but want to double-check with the maintainers.
> 
> The test fails with the previous ordering.

That should be good. thanks for taking care of the test.

**Comment by cwperks** [comment]
Merged the latest changes in from main to re-run CI

**Pull request merged**